File: cpp/weilsdk/src/weil_contracts/nonFungible.cpp
#include "weilsdk/weil_contracts/nonFungible.h"
#include <variant>
#include <vector>
#include <string>

namespace weilcontracts {

    Token::Token(std::string _title, std::string _name, std::string _description,
                std::string _payload) {
        this->title = _title;
        this->description = _description;
        this->name = _name;
        this->payload = _payload;
    }

    NonFungibleToken::NonFungibleToken(std::string _name) {
        this->name = _name;
        this->creator = weilsdk::Runtime::sender();
        this->tokens = collections::WeilMap<TokenId, Token>(1);
        this->owners = collections::WeilMap<TokenId, Address>(2);
        this->owned = collections::WeilMap<Address, std::set<TokenId>>(3);
        this->allowances = collections::WeilMap<std::string, Address>(4);
    }

    std::string NonFungibleToken::getName() const { return name; }

    void NonFungibleToken::setName(std::string _name) { name = _name;}

    std::string NonFungibleToken::getCreator() const { return creator; }

    void NonFungibleToken::setTokens(collections::WeilMap<TokenId, Token> _tokens) {
        this->tokens = _tokens;
    }

    void NonFungibleToken::setOwners(
        collections::WeilMap<TokenId, Address> _owners) {
        this->owners = _owners;
    }

    void NonFungibleToken::setOwned(
        collections::WeilMap<Address, std::set<TokenId>> _owned) {
        this->owned = _owned;
    }

    void NonFungibleToken::setAllowances(
        collections::WeilMap<std::string, Address> _allowances) {
        this->allowances = _allowances;
    }

    bool NonFungibleToken::isValidId(TokenId tokenId) {
        return tokenId.length() > 0 && tokenId.length() < 256;
    }

    bool NonFungibleToken::hasBeenMinted(TokenId tokenId) {
        collections::WeilMap<std::string, std::string> mp(this->owners.getStateId());
        bool c1 = mp.contains(tokenId);
        bool c2 = mp.get(tokenId) != EMPTY_TOKEN_ID;

        return c1 && c2;
    }

    uint32_t NonFungibleToken::balanceOf(std::string addr) {
        return this->getOwned().contains(addr) ? this->getOwned().get(addr).size()
                                            : 0;
    }

    std::pair<int, Address> NonFungibleToken::ownerOf(TokenId tokenId) {
        if (!isValidId(tokenId)) {
            return {1, tokenId + " is not a valid id"};
        }
        if (!this->getOwners().contains(tokenId)) {
            return {1, "Owner of " + tokenId + " is not identified"};
        }
        collections::WeilMap<std::string, std::string> mp(this->owners.getStateId());
        return {0, "\"" + mp.get(tokenId) + "\""};
    }

    std::pair<bool, std::variant<Token, std::string>>
    NonFungibleToken::details(TokenId tokenId) {

        if (!isValidId(tokenId)) {
            return {1, tokenId + " is not a valid id"};
        }

        if (!hasBeenMinted(tokenId)) {
            return {1, tokenId + " has not been minted yet"};
        }
        collections::WeilMap<TokenId, Token> tokens_map(
            this->getTokens().getStateId());

        if (!tokens_map.contains(tokenId)) {
            return {1, "token " + tokenId + " not found"};
        }
        return {0, tokens_map.get(tokenId)};
    }

    std::pair<bool, std::string> NonFungibleToken::doTransfer(std::string tokenId,
                                                            Address fromAddr,
                                                            Address toAddr) {
        bool result = weilsdk::Ledger::transfer(tokenId, fromAddr, toAddr, 1).first;

        if (!result) {
            return {1, tokenId + " could not be transferred by the Ledger"};
        }

        collections::WeilMap<std::string,std::string> ownersMap(this->getOwners().getStateId());
        ownersMap.insert(tokenId, toAddr);
        collections::WeilMap<std::string, std::set<TokenId>> ownedMap(this->getOwned().getStateId());
        std::set<TokenId> oldOwnersTokens = ownedMap.get(fromAddr);
        if (oldOwnersTokens.empty()) {
            return {1, "Owned tokens is missing"};
        }

        oldOwnersTokens.erase(tokenId);
        ownedMap.insert(fromAddr, oldOwnersTokens);

        std::set<TokenId> newOwnersTokens = ownedMap.get(toAddr);
        newOwnersTokens.insert(tokenId);
        ownedMap.insert(toAddr, newOwnersTokens);

        std::string key = fromAddr + "$" + tokenId;
        collections::WeilMap<std::string, std::string> allowancesMap(this->getAllowances().getStateId());
        allowancesMap.remove(key);

        return {0, "Ok"};
    }

    std::pair<bool, std::string> NonFungibleToken::transfer(Address toAddr,
                                                            TokenId tokenId) {
        Address fromAddr = weilsdk::Runtime::sender();

        collections::WeilMap<std::string, std::string> ownersMap(this->getOwners().getStateId());                                                

        if (!isValidId(tokenId)) {
            return {1, "Token " + tokenId + " is not a valid token id"};
        }

        if (!ownersMap.contains(tokenId)) {
            return {1, "Token " + tokenId + " is missing an owner"};
        }
        Address oldOwner = ownersMap.get(tokenId);
        if (oldOwner != fromAddr) {
            return {1, "Token " + tokenId + " not owned by " + fromAddr};
        }

        return doTransfer(tokenId, fromAddr, toAddr);
    }

    std::pair<bool, std::string> NonFungibleToken::transferFrom(Address fromAddr,
                                                                Address toAddr,
                                                                TokenId tokenId) {
        Address spender = weilsdk::Runtime::sender();

        if (!isValidId(tokenId)) {
            return {1, "token " + tokenId + " is not a valid token id"};
        }

        if (!this->getOwners().contains(tokenId)) {
            return {1, "token " + tokenId + " is missing an owner"};
        }

        Address oldOwner = this->getOwners().get(tokenId);
        if (oldOwner != fromAddr) {
            return {1, "token " + tokenId + " not owned by " + fromAddr};
        }

        std::string key = oldOwner + "$" + tokenId;
        bool allowed = this->getAllowances().contains(key) &&
                        this->getAllowances().get(key) == spender;

        if (!allowed) {
            std::string key2 = oldOwner + "$" + EMPTY_TOKEN_ID;
            if (!this->getAllowances().contains(key2) || this->getAllowances().get(key2) != spender){
                return {1, "transfer of token `" + tokenId + "` not authorized"};
            }
        }

        return doTransfer(tokenId, fromAddr, toAddr);
    }

    std::pair<bool, std::string> NonFungibleToken::approve(Address spender,
                                                        TokenId tokenId) {
        Address fromAddr = weilsdk::Runtime::sender();

        if (!isValidId(tokenId)) {
            return {1, "token `" + tokenId + " is not a valid token id"};
        }

        if (!this->getOwners().contains(tokenId)) {
            return {1, "token `" + tokenId + "` is missing an owner"};
        }

        Address oldOwner = this->getOwners().get(tokenId);
        if (oldOwner != fromAddr) {
            return {1, "token `" + tokenId + "` not owned by " + fromAddr};
        }

        std::string key = oldOwner + "$" + tokenId;
        if (spender == EMPTY_ADDRESS) {
            this->getAllowances().remove(key);
        } else {
            this->getAllowances().insert(key, spender);
        }

        return {0, "Ok"};
    }

    std::pair<bool, std::variant<std::string, std::vector<Address>>>
    NonFungibleToken::getApproved(TokenId tokenId) {
        std::vector<Address> response;

        if (!isValidId(tokenId)) {
            return {1, "token `" + tokenId + "` is not a valid token id"};
        }

        if (!this->getOwners().contains(tokenId)) {
            return {1, "token `" + tokenId + "` is missing an owner"};
        }

        std::string owner = this->getOwners().get(tokenId);

        std::string key1 = owner + "$" + tokenId;
        if (this->getAllowances().contains(key1)) {
            response.push_back(this->getAllowances().get(key1));
        }

        std::string key2 = owner + "$" + EMPTY_TOKEN_ID;
        if (this->getAllowances().contains(key2)) {
            response.push_back(this->getAllowances().get(key2));
        }

        return {0, response};
    }

    void NonFungibleToken::setApproveForAll(std::string spender, bool approval) {
    Address fromAddr = weilsdk::Runtime::sender();
    std::string key = fromAddr + "$" + EMPTY_TOKEN_ID;

        if (approval) {
            this->getAllowances().insert(key, spender);
        } else {
            this->getAllowances().remove(key);
        }
    }

    bool NonFungibleToken::isApprovedForAll(std::string owner,
                                            std::string spender) {
        std::string key = owner + "$" + EMPTY_TOKEN_ID;
        return this->getAllowances().contains(key) &&
               this->getAllowances().get(key) == spender;
    }

    std::pair<int, std::string> NonFungibleToken::mint(TokenId tokenId,
                                                    Token token) {
        Address sender = weilsdk::Runtime::sender();

        if (!isValidId(tokenId)) {
            return {1, "invalid token id"};
        }
        if(this->getTokens().contains(tokenId)){
            return {1, "token id `" + tokenId + "` already minted " +
                tokens.get(tokenId).name};
        }

        auto res = weilsdk::Ledger::mint(tokenId, sender, 1);
        if(!res.first){
            return {1,"could not mint through ledger"};
        }

        collections::WeilMap<TokenId, Token> tokenMap(this->getTokens().getStateId());
        collections::WeilMap<std::string, std::string> ownersMap(this->getOwners().getStateId());
        collections::WeilMap<Address, std::set<TokenId>> ownedMap(this->getOwned().getStateId());
        tokenMap.insert(tokenId, token);
        ownersMap.insert(tokenId, sender);

        if (!ownedMap.contains(sender)) {
            ownedMap.insert(sender, std::set<TokenId>());
        }

        std::set<TokenId> senderTokens = ownedMap.get(sender);
        senderTokens.insert(tokenId);
        ownedMap.insert(sender, senderTokens);
        return {0, tokenId + " has been minted by " + sender};
    }

} // namespace weilcontracts

File: adk/go/weil_go/weil_contracts/non_fungible/non_fungible.go
// Package non_fungible provides a non-fungible token (NFT) implementation for Weil contracts.
// It implements an ERC-721-like token standard with unique token IDs, ownership tracking,
// and approval mechanisms.
package non_fungible

import (
	"fmt"

	"unicode/utf8"

	"github.com/weilliptic-public/jsonmap"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/collections"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/ledger"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

// Token represents the metadata for a non-fungible token.
type Token struct {
	Title       string `json:"title"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Payload     string `json:"payload"`
}

type innerNonFungibleToken struct {
	Name       string                                   `json:"name"`
	Creator    string                                   `json:"creator"`
	Id2Tokens  collections.WeilMap[string, Token]       `json:"tokens"`
	Id2Owner   collections.WeilMap[string, string]      `json:"owners"`
	Owner2Ids  collections.WeilMap[string, jsonmap.Map] `json:"owned"`
	Allowances collections.WeilMap[string, string]      `json:"allowances"`
}

// NonFungibleToken represents a non-fungible token contract.
type NonFungibleToken struct {
	Inner innerNonFungibleToken `json:"token"`
}

// Dump returns a string representation of the internal state paths for debugging.
func (i *innerNonFungibleToken) Dump() string {
	return fmt.Sprintf("Tokens: %s, Owners: %s, Owned: %s, Allowances: %s",
		i.Id2Tokens.BaseStatePath(),
		i.Id2Owner.BaseStatePath(),
		i.Owner2Ids.BaseStatePath(),
		i.Allowances.BaseStatePath(),
	)
}

// NewNonFungibleToken creates a new non-fungible token contract with the given name.
// The creator is set to the address of the transaction sender.
func NewNonFungibleToken(name string) *NonFungibleToken {
	return &NonFungibleToken{
		Inner: innerNonFungibleToken{
			Name:       name,
			Creator:    runtime.Sender(),
			Id2Tokens:  *collections.NewWeilMap[string, Token](*collections.NewWeilId(0)),
			Id2Owner:   *collections.NewWeilMap[string, string](*collections.NewWeilId(10)),
			Owner2Ids:  *collections.NewWeilMap[string, jsonmap.Map](*collections.NewWeilId(2)),
			Allowances: *collections.NewWeilMap[string, string](*collections.NewWeilId(3)),
		},
	}
}

// Name returns the name of the NFT contract.
func (t *NonFungibleToken) Name() string {
	return t.Inner.Name
}

// Creator returns the address of the contract creator.
func (t *NonFungibleToken) Creator() string {
	return string(t.Inner.Creator)
}

// isValidId checks if a token ID is valid (non-empty and less than 255 characters).
func isValidId(tokenId string) bool {
	len := utf8.RuneCountInString(tokenId)
	return len > 0 && len < 255
}

// hasBeenMinted checks if a token with the given ID has already been minted.
func (t *NonFungibleToken) hasBeenMinted(tokenId string) bool {
	val, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return false
	}
	return *val != ""
}

// BalanceOf returns the number of tokens owned by the given address.
func (t *NonFungibleToken) BalanceOf(addr string) uint64 {
	val, err := t.Inner.Owner2Ids.Get(&addr)
	if err != nil {
		return 0
	}
	return uint64(val.Len())
}

// OwnerOf returns the address of the owner of the token with the given ID.
// Returns an error if the token ID is invalid or the token has not been minted.
func (t *NonFungibleToken) OwnerOf(tokenId string) (*string, error) {
	if !isValidId(tokenId) {
		return nil, fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	val, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return nil, fmt.Errorf("owner of `%s` is not identified", tokenId)
	}
	return val, nil
}

// Details returns the metadata for the token with the given ID.
// Returns an error if the token ID is invalid or the token has not been minted.
func (t *NonFungibleToken) Details(tokenId string) (*Token, error) {
	if !isValidId(tokenId) {
		return nil, fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	if !t.hasBeenMinted(tokenId) {
		return nil, fmt.Errorf("token `%s` has not been minted", tokenId)
	}

	val, err := t.Inner.Id2Tokens.Get(&tokenId)
	if err != nil {
		return nil, fmt.Errorf("token `%s` not found", tokenId)
	}
	return val, nil
}

// doTransfer performs the internal transfer logic for moving a token from one address to another.
// This is a helper function used by Transfer and TransferFrom.
func (t *NonFungibleToken) doTransfer(tokenId string, fromAddr string, toAddr string) error {
	// Update the Ledger
	err := ledger.Transfer(tokenId, fromAddr, toAddr, 1)
	if err != nil {
		return fmt.Errorf("`%s` could not be transferred by the Ledger", tokenId)
	}

	// Update the token
	t.Inner.Id2Owner.Insert(&tokenId, &toAddr)

	// Update the owners
	oldOwnersTokens, err := t.Inner.Owner2Ids.Get(&fromAddr)
	if err != nil {
		return fmt.Errorf("owned token is missing")
	}
	oldOwnersTokens.Delete(tokenId)
	t.Inner.Owner2Ids.Insert(&fromAddr, oldOwnersTokens)

	newOwnersTokens, err := t.Inner.Owner2Ids.Get(&toAddr)
	if err != nil {
		newOwnersTokens = jsonmap.New()
	}
	newOwnersTokens.Set(tokenId, nil)
	t.Inner.Owner2Ids.Insert(&toAddr, newOwnersTokens)

	// Update old owner's individual allowances
	key := fmt.Sprintf("%s$%s", fromAddr, tokenId)
	_, _ = t.Inner.Allowances.Remove(&key)
	return nil
}

// Transfer transfers a token from the sender's address to the specified address.
// Returns an error if the token ID is invalid, the token is not owned by the sender, or the transfer fails.
func (t *NonFungibleToken) Transfer(toAddr string, tokenId string) error {
	fromAddr := runtime.Sender()

	if !isValidId(tokenId) {
		return fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	oldOwner, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return fmt.Errorf("owner of `%s` is not identified", tokenId)
	}

	if *oldOwner != fromAddr {
		return fmt.Errorf("token `%s` now owned by `%s`", tokenId, fromAddr)

	}
	return t.doTransfer(tokenId, fromAddr, toAddr)
}

// TransferFrom transfers a token from one address to another on behalf of the sender.
// The sender must be the owner or have been approved to transfer the token.
// Returns an error if the transfer is not authorized or fails.
func (t *NonFungibleToken) TransferFrom(fromAddr string, toAddr string, tokenId string) error {
	spender := runtime.Sender()

	if !isValidId(tokenId) {
		return fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	oldOwner, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return fmt.Errorf("owner of `%s` is not identified", tokenId)
	}

	if *oldOwner != fromAddr {
		return fmt.Errorf("token `%s` not owned by `%s`", tokenId, fromAddr)
	}

	key := fmt.Sprintf("%s$%s", *oldOwner, tokenId)

	var allowed bool
	allowance, err := t.Inner.Allowances.Get(&key)
	if err != nil {
		allowed = false
	} else {
		allowed = *allowance == spender
	}

	if !allowed {
		key = fmt.Sprintf("%s$", *oldOwner)

		allowance, err = t.Inner.Allowances.Get(&key)
		if err != nil {
			allowed = false
		} else {
			allowed = *allowance == spender
		}
	}

	if !allowed {
		return fmt.Errorf("transfer of token `%s` not authorized", tokenId)
	}

	return t.doTransfer(tokenId, fromAddr, toAddr)
}

// Approve approves the spender to transfer the specified token.
// Only the owner of the token can approve a spender.
// If spender is an empty string, the approval is removed.
// Returns an error if the token ID is invalid or the sender is not the owner.
func (t *NonFungibleToken) Approve(spender string, tokenId string) error {
	fromAddr := runtime.Sender()

	if !isValidId(tokenId) {
		return fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	oldOwner, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return fmt.Errorf("owner of `%s` is not identified", tokenId)
	}

	if *oldOwner != fromAddr {
		return fmt.Errorf("allowance of token `%s` not authorized", tokenId)
	}

	key := fmt.Sprintf("%s$%s", fromAddr, tokenId)
	if spender == "" {
		t.Inner.Allowances.Remove(&key)
	} else {
		t.Inner.Allowances.Insert(&key, &spender)
	}
	return nil
}

// SetApproveForAll approves or revokes approval for the spender to transfer all tokens
// owned by the sender. If approval is true, the spender is approved; if false, approval is revoked.
func (t *NonFungibleToken) SetApproveForAll(spender string, approval bool) {
	fromAddr := runtime.Sender()

	key := fmt.Sprintf("%s$", fromAddr)
	if approval {
		t.Inner.Allowances.Insert(&key, &spender)
	} else {
		t.Inner.Allowances.Remove(&key)
	}
}

// GetApproved returns a list of approved spenders for the given token.
// This includes both token-specific approvals and operator approvals (approve for all).
// Returns an error if the token ID is invalid or the token has not been minted.
func (t *NonFungibleToken) GetApproved(tokenId string) (*[]string, error) {
	response := make([]string, 0)
	if !isValidId(tokenId) {
		return nil, fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	owner, err := t.Inner.Id2Owner.Get(&tokenId)
	if err != nil {
		return nil, fmt.Errorf("owner of `%s` is not identified", tokenId)
	}

	key := fmt.Sprintf("%s$%s", *owner, tokenId)
	allowance, err := t.Inner.Allowances.Get(&key)
	if err == nil {
		response = append(response, *allowance)
	}
	key = fmt.Sprintf("%s$", *owner)
	allowance, err = t.Inner.Allowances.Get(&key)
	if err == nil {
		response = append(response, *allowance)
	}
	return &response, nil
}

// IsApprovedForAll checks if the spender is approved to transfer all tokens owned by the owner.
func (t *NonFungibleToken) IsApprovedForAll(owner string, spender string) bool {
	key := fmt.Sprintf("%s$", owner)

	allowance, err := t.Inner.Allowances.Get(&key)
	if err != nil {
		return false
	}
	return *allowance == spender
}

// Mint creates a new token with the given ID and metadata, assigning it to the sender.
// Returns an error if the token ID is invalid, the token has already been minted, or minting fails.
func (t *NonFungibleToken) Mint(tokenId string, token Token) error {
	fromAddr := runtime.Sender()

	if !isValidId(tokenId) {
		return fmt.Errorf("`%s` is not a valid token id", tokenId)
	}

	if t.hasBeenMinted(tokenId) {
		return fmt.Errorf("token `%s` has already been minted", tokenId)
	}

	err := ledger.Mint(tokenId, fromAddr, 1)
	if err != nil {
		return fmt.Errorf("`%s` could not be transferred by the Ledger", tokenId)
	}

	t.Inner.Id2Tokens.Insert(&tokenId, &token)
	t.Inner.Id2Owner.Insert(&tokenId, &fromAddr)

	newOwnersTokens, err := t.Inner.Owner2Ids.Get(&fromAddr)
	if err != nil {
		newOwnersTokens = jsonmap.New()
	}

	newOwnersTokens.Set(tokenId, nil)
	t.Inner.Owner2Ids.Insert(&fromAddr, newOwnersTokens)
	return nil
}
